index.d.ts

```diff
@@ -734,21 +734,15 @@ interface ExtendedStorageInstance {
    * @param dbPath - Database path to place uploaded file metadata
    * @param options - Options
    * @param options.name - Name of the file
+   * @param options.metadata - Metadata associated with the file to upload to storage
+   * @param options.documentId - Id of document to update with metadata if using Firestore
    * @see https://react-redux-firebase.com/docs/api/storage.html#uploadFile
    */
   uploadFile: (
     path: string,
-    file: File,
+    file: File | Blob,
```

Comment:
Glad you caught this!

Reply:
@prescottprue @JonnyBoy333 

Adding `Blob` is now breaking my type check:

```typescript
  const { uploadTaskSnapshot } = await firebase.uploadFile(
          storagePath,
          acceptedFiles[0],
          undefined,
          {
            name: (file: File) => file.name.replace(/[^\w.\-_]/gi, ''),
          }
        );
```

```
Type 'File | Blob' is not assignable to type 'File'.
            Type 'Blob' is missing the following properties from type 'File': lastModified, name ts(2345)
```

This is because the new type definition is incorrect:

```typescript
export interface UploadFileOptions {
  name?: string | ((
    file: File | Blob, // should not use the "|" operator
    internalFirebase: WithFirebaseProps<ProfileType>['firebase'],
    uploadConfig: {
      path: string,
      file: Either<File, Blob>,
      dbPath?: string,
      options?: UploadFileOptions
    }
  ) => string)
  ...
}
```

```typescript
export interface UploadFileOptions {
  name?: string | ((
    file: Either<File, Blob>, // use Either<type1, type2> instead
    internalFirebase: WithFirebaseProps<ProfileType>['firebase'],
    uploadConfig: {
      path: string,
      file: Either<File, Blob>,
      dbPath?: string,
      options?: UploadFileOptions
    }
  ) => string)
  ...
}
```

Reply:
Interesting, I just tried recreating your code and I am not getting any typescript errors. The fact that you are explicitly setting the file to a File type with `file: File` should let typescript know that it is a File and not a Blob, which allows access to the name property (I do the same thing in my code except for Blob where the name property does not exist).

``` typescript
    const testFile = new File([], 'test');
    const { uploadTaskSnapshot } = await firebase.uploadFile(
      'test/path',
      testFile,
      undefined,
      {
        name: (file: File) => file.name.replace(/[^\w.\-_]/gi, ''),
      }
    );
    // No typescript errors here
```

The reason for the union type here is because firebases file storage api does support both File and Blob types (as well as Uint8Array). And what type gets passed into the name function is dependent on the type of the second property of the `uploadFile` function.

There might be a better way to tell typescript that since a File type object was put in the `uploadFile` function then it can be certain that a File type object will get passed to the `name` function. I'd have to think about that a little more.

Reply:
Thanks for reaching out - So guessing that should have been `Either<File, Blob>`?

Reply:
@prescottprue @JonnyBoy333 Created a PR with the fix here: #950

Reply:
@rscotten That pull request is what I was thinking of but didn't know how to do myself. Thank you! Now I'm guessing you would write your upload that like this?

``` typescript
const testFile = new File([], 'test');
const { uploadTaskSnapshot } = await firebase.uploadFile<File>( // Adding the type here
  'test/path',
  testFile,
  undefined,
  {
    name: (file) => file.name.replace(/[^\w.\-_]/gi, ''),
  }
);
```

Reply:
@JonnyBoy333 yep! That's correct.